internal/overlord/logstate/syslog.go

```diff
@@ -0,0 +1,184 @@
+	version        int
+	host           string
+	msgid          string
+	priority       int
+	pid            string
+	structuredData string // TODO: use this for future log labels
+
+	// mu helps manage access to network connection-related internal state.  This should be
+	// read-locked whenever using the syslog server connection and write-locked whenever
+	// replacing/reconnecting/etc the syslog server connection.
+	mu            sync.RWMutex
+	conn          net.Conn
+	address       *url.URL
+	waitReconnect time.Duration
+	closed        bool
+}
+
+// NewSyslogBackend creates a writer forwarding writes as syslog messages to dst.  The forwarded
+// messages will have app as the application name.  Other message parameters are set using
+// reasonable defaults or the RFC5424 nil value "-".
+func NewSyslogBackend(addr string) (*SyslogBackend, error) {
+	u, err := url.Parse(addr)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + addr)
+	}
+	if err != nil {
+		return nil, err
+	} else if u.Scheme != "tcp" && u.Scheme != "udp" && u.Scheme != "" {
+		return nil, fmt.Errorf("invalid syslog server address scheme %q", u.Scheme)
+	}
+	if u.Scheme == "" {
+		u.Scheme = "tcp"
+	}
+
+	return &SyslogBackend{
+		version:        1,
+		pid:            "-",
+		host:           "-",
+		msgid:          "-",
+		priority:       1*8 + 6, // for facility=user-msg severity=informational. See RFC 5424 6.2.1 for available codes.
+		structuredData: syslogStructuredData("pebble", canonicalPrivEnterpriseNum, nil),
+
+		address: u,
+	}, nil
+}
+
+// syslogStructuredData formats the given labels into a structured data section for a syslog message
+// according to RFC5424 section 6.
+func syslogStructuredData(name string, enterpriseNum int, labels map[string]string) string {
+	if len(labels) == 0 {
+		return "-"
+	}
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "[%s@%d", name, enterpriseNum)
+	for key, value := range labels {
+		fmt.Fprintf(&buf, " %s=\"", key)
+		// escape the value according to RFC5424 6.3.3
+		for i := 0; i < len(value); i++ {
+			// don't use "for _, c := range value" as we don't want runes
+			c := value[i]
+			if c == '"' || c == '\\' || c == ']' {
+				buf.WriteByte('\\')
+			}
+			buf.WriteByte(c)
+		}
+		buf.WriteByte('"')
+	}
+	buf.WriteByte(']')
+	return buf.String()
+}
+
+func (s *SyslogBackend) Send(m *LogMessage) error {
+	err := s.ensureConnected()
+	if err != nil {
+		return err
+	}
+
+	s.mu.RLock()
+	_, err = s.conn.Write(s.buildMsg(m))
+	s.mu.RUnlock()
+
+	if err != nil {
+		// The connection might be bad. Close and reset it for later reconnection attempt(s).
+		s.mu.Lock()
+		s.conn.Close()
+		s.mu.Unlock()
+	}
+	return err
+}
+
+func (s *SyslogBackend) buildMsg(m *LogMessage) []byte {
+	// format defined by RFC 5424
+	timestamp := m.Timestamp.Format(time.RFC3339)
+	msg := fmt.Sprintf("<%d>%d %s %s %s %s %s %s %s",
+		s.priority, s.version, timestamp, s.host, m.Service, s.pid, s.msgid, s.structuredData, m.Message)
```

Comment:
explicitly write the Message and structuredData to avoid so many allocations and copies